src/python/pants/engine/addressable.py

```diff
@@ -32,9 +31,8 @@ def expect_single(self) -> Address:
         return self.dependencies[0]
 
 
-@dataclass(frozen=True)
-class AddressWithOrigin:
-    """A BuildFileAddress along with the cmd-line spec it was generated from."""
+class AddressWithOrigin(NamedTuple):
```

Comment:
So, the reason the `datatype` subclass came into existence in the first place is that NamedTuple does not account for type differences in `eq` and `hash`. So any other class with the same fields is potentially "equal" to this one (and one time, early in the development of the engine, I spent about 7 hours trying to track down a weird failure because of that, heh).

That's probably unlikely in this case given the verbosity of the name, but it is likely if more usage of `NamedTuple` is triggered by this change.

So: should we reintroduce `datatype`?

Reply:
Indeed on `__eq__` and `__hash__`. I figured it's unlikely we'll ever have another representation of `(Address, OriginSpec)` that gets used with the engine.

> should we reintroduce datatype?

I don't think so. This was a really useful abstraction before dataclasses, but it was a lot of custom code and MyPy didn't understand it. 

If we think `__eq__` and `__hash__` is too much of a concern to land this, then I'd rather close this PR and stick with dataclasses.

Reply:
If you go back far enough, `datatype` was pretty simple: https://github.com/pantsbuild/pants/blob/517fc382bbc22320480a0e13e69c2e63a0f94a2a/src/python/pants/util/objects.py

It's not clear to me why datatype wouldn't work with mypy in its original form. Constructing via a method call was just to imitate `namedtuple` syntax at the time.

Reply:
Or even further, before methods were added to ban iteration: https://github.com/pantsbuild/pants/blob/823bc0494ae30f6eb62015662af225369fb25fdc/src/python/pants/util/objects.py